### PR TITLE
Add `type` property to `FunctionParameter` and `ReturnValue`

### PR DESCRIPTION
**Pull request by Hell-Tractor**
This pr is similar to #687.

At present, `FunctionParameter` only contains one property `name`. As I know, analyzing function signature is the only way to get parameter types and return value type.

Therefore these changes were made to support querying function parameter types and return value type directly. The new structure of `FunctionParameter` and `ReturnValue` is:

```graphql
type FunctionParameter {
    name: String!
    type: String!
}
type ReturnValue {
    is_unit: Boolean!
    type: String!
}
```

To make stringified `type` as similar to origin code as possible, it is converted from `rustdoc_types::Type` to `String` by using function `adapter::rust_type_name::rust_type_name`. The property `type` of function without return value is set to `()`.

### CODE REVIEW

File: test_crates/function_param_types_and_return_value_types/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+pub fn fn_with_generic_ref<T: std::ops::Add<Output = T>>(a: &T, b: &T) -> T
+where
+    T: Copy,
+{
+    *a + *b
+}
+
+pub fn fn_with_array_param(arr: [i32; 3]) -> i32 {
+    arr.iter().sum()
+}
```
Comment (obi1kenobi) [inline test_crates/function_param_types_and_return_value_types/src/lib.rs:35]: Putting a single trailing newline at the end of files makes other tools work better.
```suggestion
}

```

If you use VSCode or a derived editor, I recommend adding the following to your config:
```
    "files.trimFinalNewlines": true,
    "files.trimTrailingWhitespace": true,
    "files.insertFinalNewline": true,
```

File: src/adapter/tests.rs [hunk 2]
```diff
@@ -8182,3 +8182,92 @@ fn function_return_value() {
 
     similar_asserts::assert_eq!(expected_results, results);
 }
+
+#[test]
+fn function_parameter_types_and_return_types() {
+    get_test_data!(data, function_param_types_and_return_value_types);
```
Comment (obi1kenobi) [inline src/adapter/tests.rs:8188]: This test is good. Would you mind adding another, separate test function that uses the `raw_type_json` test crate with this same query?

It has some much more complex types and we want to make sure every flavor of type we can think of is thoroughly covered by tests.

File: src/adapter/properties.rs [hunk 3]
```diff
@@ -377,6 +385,15 @@ pub(super) fn resolve_return_value_property<'a, V: AsVertex<Vertex<'a>> + 'a>(
...
                 .into()
         }),
+        "type" => resolve_property_with(contexts, |vertex| {
+            vertex
+                .as_return_value()
+                .expect("not a return value")
+                .type_
+                .map(rust_type_name::rust_type_name)
+                .unwrap_or("()".to_string())
+                .into()
```
Comment (obi1kenobi) [inline src/adapter/properties.rs]: Let's try to avoid this `to_string()` call, because it'll be relatively common and we'll end up allocating twice (`.into()` will allocate again).

```suggestion
                .map(rust_type_name::rust_type_name)
                .map(Into::into)
                .unwrap_or("()".into())
```
Something like this should do the trick.

File: src/rustdoc_schema.graphql [hunk 4]
```diff
@@ -1496,6 +1496,7 @@ https://docs.rs/rustdoc-types/0.11.0/rustdoc_types/struct.FnDecl.html
 """
 type FunctionParameter {
   name: String!
+  type: String!
```
Comment (obi1kenobi) [inline src/rustdoc_schema.graphql:1542]: Is using `type` as an identifier syntactically valid here? Wow, I wouldn't have expected that. Nice!

Please make sure to add a thorough doc comment above it though? Unlike `name`, there are _many_ edge cases we need to write down here. Look at the `signature` doc comment as a starting point, though `type` certainly has more things to mention. For example:
- we do not specify whether types contained within use qualified paths or just names, nor a particular qualified path over another (e.g. `String` vs `std::string::String` vs `alloc::string::String`)
- generic parameters will appear without being defined and without their bounds, since the definition and bounds are syntactically part of the function, not the parameter

etc.

I recommend reading through the `raw_type_json` crate for some ideas how complex parameter types can get, and then writing down more docs :)

File: src/rustdoc_schema.graphql [hunk 5]
```diff
@@ -2007,6 +2008,8 @@ type ReturnValue {
   ```
   """
   is_unit: Boolean!
+
+  type: String!
```
Comment (obi1kenobi) [inline src/rustdoc_schema.graphql:2098]: We'll want a nice thorough doc comment here too.

**Review by obi1kenobi — COMMENTED** [review]
Hey, thanks for the PR! Just a handful of suggestions about improving docs and testing and this should be good to merge.

If you don't mind me asking, I'm curious how you found this project and were inspired to open a PR! Are you already using it somewhere, or did it just seem interesting?

If you'd like to join the project's Discord, you're welcome to: https://discord.gg/wqhMJYfa

**Comment by Hell-Tractor** [comment]
> If you don't mind me asking, I'm curious how you found this project and were inspired to open a PR! Are you already using it somewhere, or did it just seem interesting?

I'm a Master's students and my team work on detecting vulnerabilities on rust programs. My teammate found this project and this project is quite helpful for analyzing api docs of rust crates. We're trying using this project as part of our detection method. Our method will need type information of function input params and return value, so I tried to augment this project and this pr was made.

I have joined the Discord group and thank you for inviting me.

I'm now working on new  test cases and doc comment. Changes will be made later.

File: src/adapter/tests.rs (hunk 2 above)
Comment (Hell-Tractor) [reply src/adapter/tests.rs:8188]: Test function was add through commit f6e333c and 8c63736. The code seems working fine.

One thing to be noted:

In `raw_type_json/src/lib.rs`, parameter `d` in function `my_generic_function` is declared as:

```rust
d: impl for<'x> FnMut(
            &'a unsafe extern "C" fn(
                *const [u8],
                &'x mut *mut (),
                ...
            ) -> std::borrow::Cow<'static, [u8]>,
        ) -> &'x (dyn std::fmt::Display)
        + Send
        + 'static,
```

Stringified type of `d` is (I have added some white spaces manually to make it more readable):

```rust
impl for<'x> FnMut(
    &'a unsafe extern "C" fn(
        _: *const [u8],
        _: &'x mut *mut (),
        ...
    ) -> std::borrow::Cow<'static, [u8]>
) -> &'x (dyn std::fmt::Display)
+ Send
+ 'static
```

It's easy to notice that parameters of `unsafe extern "C" fn` was automatically added `_` as param names. Is this behavior expected?

File: src/rustdoc_schema.graphql (hunk 4 above)
Comment (Hell-Tractor) [reply src/rustdoc_schema.graphql:1542]: I have added some doc comments through commit 957c0ae and 73e55be, specifying how type paths, generic parameters, life time parameters and `Self` type was stringified.

Is there anything more to be specified?

File: src/rustdoc_schema.graphql (hunk 5 above)
Comment (Hell-Tractor) [reply src/rustdoc_schema.graphql:2098]: The doc comment here is almost same with the one in `FunctionParameter`. An additional note is included here stating that a unit return value will be stringified as `()`.

File: src/adapter/tests.rs (hunk 2 above)
Comment (obi1kenobi) [reply src/adapter/tests.rs:8188]: Is that syntax valid? I'd be surprised if so.

If it isn't valid syntax, we should probably fix it and make it emit correct syntax for function pointer types.

File: src/rustdoc_schema.graphql [hunk 6]
```diff
@@ -1496,6 +1496,50 @@ https://docs.rs/rustdoc-types/0.11.0/rustdoc_types/struct.FnDecl.html
 """
 type FunctionParameter {
   name: String!
+
+  """
+  Stringified representation of the parameter's type, as close to the
+  original source code as possible.  May not preserve whitespace, constant
```
Comment (obi1kenobi) [inline src/rustdoc_schema.graphql:1502]: nit: the docs style is to use single space between sentences
```suggestion
  original source code as possible. May not preserve whitespace, constant
```

File: src/rustdoc_schema.graphql [hunk 7]
```diff
@@ -1496,6 +1496,50 @@ https://docs.rs/rustdoc-types/0.11.0/rustdoc_types/struct.FnDecl.html
...
+  expressions in types, or path names, but the output will be semantically
+  equivalent to the original.
+
+  The stringified type always within use full qualified path.
```
Comment (obi1kenobi) [inline src/rustdoc_schema.graphql:1506]: "within" seems like a typo here.

Also, are you sure we can guarantee this in practice? If so, we'll need *a lot* more tests and documentation. Some of the edge cases we'll need to discuss:
- What if the type is defined at some path that is not publicly accessible, but it's re-exported at some other publicly importable path? Which fully-qualified path will be used?
- What if there's more than one fully-qualified path that is publicly importable, which one will be used? Concretely, why is `std::option::Option` chosen and not [`core::option::Option`](https://doc.rust-lang.org/core/option/enum.Option.html)?
- What if the type is defined in one crate and re-exported in another one, which fully-qualified path would we use? Does it matter if our crate has dependencies on both or only one of those crates where the type is defined or re-exported?

etc. etc.

This seems like a lot of work to define, test, and get merge-ready. In particular, we can't reasonably even say "we output whichever path rustdoc JSON gives us" because rustdoc JSON is an unstable output format — it definitely doesn't specify what path comes back, and it in principle could be different between different nightlies, on different platforms, or even between compilations on the same platform. So if we promise something stricter, we risk not being able to uphold it without a lot more work.

It may be better to offer less strict guarantees — but ones we can clearly define and which we know we can always uphold. For example, we could say it isn't well-defined whether fully-qualified or relative paths are used, or which path is used when there's more than one possibility. We can promise the shown path will be syntactically valid Rust, but not necessarily `pub`, not necessarily reachable from any particular location (not the same thing as `pub`), not necessarily shortest, not necessarily from the crate where the item was defined. It's _a_ way to refer to the item, but not any _specific_ way.

File: src/rustdoc_schema.graphql [hunk 8]
```diff
@@ -1496,6 +1496,50 @@ https://docs.rs/rustdoc-types/0.11.0/rustdoc_types/struct.FnDecl.html
...
+  ```
+  fn generic_fn<T: Copy>(param: T) {}
+  ```
+
+  Lifetime parameters will be preserved.
+
+  For example, in the following code, the parameter type string is `&'a str`:
+  ```
+  fn lifetime_fn<'a>(param: &'a str) {}
+  ```
```
Comment (obi1kenobi) [inline src/rustdoc_schema.graphql:1531]: How about in lifetime bounds like `+ '_`? These are common e.g. when returning iterators that borrow from `&self` but return copied/cloned values: the _iterator_ has lifetime `'_` ("the self lifetime") but the items it produces are owned.

File: src/rustdoc_schema.graphql [hunk 9]
```diff
@@ -1496,6 +1496,50 @@ https://docs.rs/rustdoc-types/0.11.0/rustdoc_types/struct.FnDecl.html
...
+
+  `Self` types in methods will appear as `Self`.
+
+  For example, in the following code, the parameter type string is `Self`:
```
Comment (obi1kenobi) [inline src/rustdoc_schema.graphql:1535]: It's a bit awkward to say "parameter type string" because there's risk of confusion with `String`, the type that a parameter can have.

How about we instead comprehensively rename `type` in the name of the property to `type_signature` (which is more clearly represented as a string) and then say "signature" here:
```suggestion
  For example, in the following code, the parameter type signature is `Self`:
```

File: src/rustdoc_schema.graphql [hunk 10]
```diff
@@ -1496,6 +1496,50 @@ https://docs.rs/rustdoc-types/0.11.0/rustdoc_types/struct.FnDecl.html
...
+  ```
+  impl MyStruct {
+      fn method(self) {}
+  }
+  ```
+  """
+  type: String!
```
Comment (obi1kenobi) [inline src/rustdoc_schema.graphql:1542]: Per my suggestion above, let's consider renaming this to `type_signature`?
```suggestion
  type_signature: String!
```

This would allow us to (in the future) add a `type` edge with more details (e.g. "is the type `Copy`" which has syntactic and SemVer implications).

Or if you feel like doing an A+ job in this PR, we could make that `type` edge immediately, and then add a new vertex type called `Type` with a property `signature` where all this would happen. Your call! If you don't feel like doing it right away, I'll probably refactor to this before too long.

File: test_crates/function_param_types_and_return_value_types/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+pub fn fn_with_qualified_param(p: std::option::Option<i32>) -> i32 {
+    p.unwrap_or(0)
+}
+
+use std::option::Option;
+pub fn fn_with_unqualified_param(p: Option<i32>) -> i32 {
+    p.unwrap_or(0)
+}
```
Comment (obi1kenobi) [inline test_crates/function_param_types_and_return_value_types/src/lib.rs:35]: nit: we always leave a trailing line at the end of files, to improve interoperability with terminal tools (there's an IDE setting for this in all IDEs if you want to not check it manually)
```suggestion
}

```

File: src/rustdoc_schema.graphql (hunk 4 above)
Comment (obi1kenobi) [reply src/rustdoc_schema.graphql:1542]: I unfortunately think there's more to be specified, and some things might be better if *explicitly* stated to not be guaranteed to happen one way or another. I left some more comments inline.

**Review by obi1kenobi — COMMENTED** [review]
Thanks again for doing this work, and nice find on the bug with function pointer parameter types! That's definitely worth fixing, possibly in a separate PR if you feel like it.

Unfortunately there's more work on the doc comments and edge cases for writing out type signatures. I'm hoping it's one more iteration on the PR, though it might be more.

Also, this PR is pretty complex so I strongly recommend re-reviewing it for typos and flow after making a commit. It's easy for things to sneak by, and generally much easier to find those things when looking at the GitHub PR interface than one's own editor. Doing that will be a huge help for me, and will also help us both merge this faster.

**Comment by obi1kenobi** [comment]
Hi! Just checking if you're still interested in driving this PR forward, or if you'd prefer if I get it over the finish line. I'm in no rush, just wanted to see if you still had the time and interest or if you got busy with other things :)

**Comment by Hell-Tractor** [comment]
I'm busy on other works these days. I will be free in about a month. If you're in a hurry, then I'll have to trouble you to complete the rest of the parts.

**Comment by obi1kenobi** [comment]
I'm happy to wait, just making sure you'll have the time and interest to
continue!

Enjoy the holidays :)

**Comment by Hell-Tractor** [comment]
Same to you!

Merry Christmas XD

**Comment by obi1kenobi** [comment]
Mostly superseded by #1064, so closing this. If you might still be interested in contributing, I'd love the help — you're welcome back anytime :)
